src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3600,34 +3634,200 @@ int RGWRados::rewrite_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, c
                        attrset, 0, real_time(), NULL, dpp, y);
 }
 
-int RGWRados::reindex_obj(const RGWBucketInfo& bucket_info,
+
+int RGWRados::reindex_obj(rgw::sal::Driver* driver,
+			  RGWBucketInfo& bucket_info, // NB: could this be made const?
 			  const rgw_obj& obj,
 			  const DoutPrefixProvider* dpp,
 			  optional_yield y)
 {
-  if (bucket_info.versioned()) {
-    ldpp_dout(dpp, 10) << "WARNING: " << __func__ <<
-      ": cannot process versioned bucket \"" <<
-      bucket_info.bucket.get_key() << "\"" <<
-      dendl;
-    return -ENOTSUP;
+  int ret;
+  RGWObjectCtx obj_ctx(driver);
+  RGWRados::Object rados_obj(this, bucket_info, obj_ctx, obj);
+
+  RGWRados::Object::Stat stat_op(&rados_obj);
+  ret = stat_op.stat_async(dpp);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": unable to initiate async stat of \"" << obj << "\"; returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
   }
 
-  Bucket target(this, bucket_info);
-  RGWRados::Bucket::UpdateIndex update_idx(&target, obj);
-  const std::string* no_write_tag = nullptr;
+  librados::IoCtx head_obj_ctx;
+  ret = get_obj_head_ioctx(dpp, bucket_info, obj, &head_obj_ctx);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": get_obj_head_ioctx for \"" << obj << "\" returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
+  }
 
-  int ret = update_idx.prepare(dpp, RGWModifyOp::CLS_RGW_OP_ADD, no_write_tag, y);
+  const uint64_t epoch = head_obj_ctx.get_last_version();
+  const int64_t poolid = head_obj_ctx.get_id();
+
+  const bool is_versioned = bucket_info.versioned();
+  const bool has_instance = ! obj.key.instance.empty();
+
+  ldpp_dout(dpp, 20) << "INFO: " << __func__ << ": reindexing " <<
+    obj.key << " in bucket " << bucket_info.bucket << dendl;
+
+  // complete stat op
+  ret = stat_op.wait(dpp);
+  if (ret == -ENOENT && is_versioned && has_instance) {
+    // head object does not exist if it's a delete marker; handle here
+    // and return
+    ldpp_dout(dpp, 20) << "INFO: " << __func__ << ": indexing " <<
+      obj.key << " as delete marker" << dendl;
```

Comment:
i'm having trouble following the logic here. we're calling `reindex_obj()` on a specific object version. why should `ENOENT` mean that we need a delete marker?

if we do need to create a delete marker, we'd learn that when we process the olh head object. i see that your `rgw-restore-bucket-index` script recognizes those:
```bash
	# we currently don't need the delete marker, but we can have access to it
	# delete_marker=$(jq -r ".removed" $olh_info_json) # true or false
```
maybe i'm misunderstanding the interaction between the script and radosgw-admin?

Reply:
i guess this logic corresponds to this part at the end?
```
 	echo "$o	$last_instance" # now add the final instance; could be delete marker; note TAB
```

Reply:
Yeah, this is one I discovered. So I have test which for a specific versioned bucket performs the following ops on it:

```
s3cmd put user-config.jam s3://$b2/obj2                                                                                                                             
s3cmd put ceph.conf s3://$b2/obj2                                                                                                                                   
s3cmd rm s3://$b2/obj2 
```

When I then do a `rados ls` of the data pool and filter for the marker for that bucket I get:

```
e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3_obj2
e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3__:gV7Jpsb8Ka7UozEbBDfcnRSYZe4lKtA_obj2
e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3__:T4JqwhU9dm3SX3TY0Ohl30vNDNstjk5_obj2
```

As you can see we have the OLH object and two instance versions. Let's examine the contents of the OLH object:

```
rados -p default.rgw.buckets.data getxattr e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3_obj2 user.rgw.olh.info >foo1
ceph-dencoder import foo1 type RGWOLHInfo decode dump_json
```

The output is:
```
{
    "target": {
        "bucket": {
            "name": "bkt2",
            "marker": "e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3",
            "bucket_id": "e32a846b-de49-4270-a8d4-c9f7d4e60c5c.14111.3",
            "tenant": "",
            "explicit_placement": {
                "data_pool": "",
                "data_extra_pool": "",
                "index_pool": ""
            }
        },
        "key": {
            "name": "obj2",
            "instance": "kwBjCsfmMm8RZCwYbi5Flb..uKEtNie",
            "ns": ""
        }
    },
    "removed": true
}
```

It's a delete marker since the `removed` field is true. Further the instance ID (`kwBjCsfmMm8RZCwYbi5Flb..uKEtNie`) is not reflected among the rados objects.

So as you noticed, the instance in the OLH object is the last instance ID presented to the code. It will not be able to find a head object that corresponds to it, so we can deduce it's a delete marker. I didn't do a complete survey of the code that creates delete markers, but it appears that it does not create a head object for delete markers. That information is simply stored in the OLH.

From previous issues we've dealt with, a series of deletes of one object used to create a sequence of delete markers in the bucket index (we've now stopped doing that), but not header objects (or any object) in the data pool to reflect them. Since we're regenerating the index from the header objects, there's no way for us to know about those other delete markers and therefore we cannot put them in the bucket index. But functionally this seems to be relatively good.